Use postcss-selector-parser for dePseudify()

corpulentcoffee opened the pull request:
Additionally, adds the `:focus-within` pseudo class and adds support for stripping pseudos even if they aren't all lowercase. Includes some tests that would have failed under the old "entire line at once" regex method.

Fixes #349.
Fixes #351.

RyanZim commented on tests/depseudify.js:
```diff
@@ -0,0 +1,89 @@
+    ], [
+        'p:not(.fancy):hover',
+        'p:not(.fancy)'
+    ]].
```
This huge map of fixtures would be cleaner as an object:

```js
{
  'p:not(.fancy):hover': 'p:not(.fancy)',
  ...
}
```

corpulentcoffee replied:
I agree! I'll change it.

RyanZim commented on src/lib.js:
```diff
@@ -28,11 +29,22 @@ const dePseudify = (function () {
-    return function (selector) {
-        return selector.replace(pseudosRegex, '$1');
+        // Actual regex is of the format: /^(:hover|:focus|...)$/i
+        pseudosRegex = new RegExp('^(' + ignoredPseudos.join('|') + ')$', 'i');
```
Isn't this indented too far?

corpulentcoffee replied:
I actually didn't change the indentation of that (it's still in the 9th column after 8 spaces), but I can see how it looks weird followed by the new `function` (which is required by the `func-style` in-use by the project; I can't make it another `const` and have the linting pass) and `const` definition after that (which uses the function).

I could give `pseudosRegex` its own `const`, but *then* eslint wants me to dedent the entire `ignoredPseudos` block since it would then by itself, and that's a bigger diff.

I can change the formatting here to whatever you'd like, though, if you don't mind the bigger diff.

RyanZim replied:
NVM, git just collapsed the diff at weird spots and made this look wrong. The indentation is perfectly fine.

RyanZim commented:
Also, just to verify, did you actually test that with HTML `<div></div>`, and CSS `div:focus-within { background: red; }`, the CSS isn't stripped?

corpulentcoffee commented:
> Also, just to verify, did you actually test that with HTML `<div></div>`, and CSS `div:focus-within { background: red; }`, the CSS isn't stripped?

Yes, I checked that out to make sure it would fix #349:

```
$ cat > test.html
<div></div>
^D

$ git checkout master 
Switched to branch 'master'
Your branch is up to date with 'upstream/master'.

$ ./bin/uncss --raw 'div:focus-within { background: red; }' test.html 


$ git checkout depseudify-via-psp 
Switched to branch 'depseudify-via-psp'
Your branch is up to date with 'fork/depseudify-via-psp'.

$ ./bin/uncss --raw 'div:focus-within { background: red; }' test.html 
div:focus-within { background: red; }
```

(This specific example is kind of silly, though, since one could argue it *should* strip the CSS rule since `<div></div>` has no child elements and therefore `:focus-within` would never match, but we don't have the machinery to do something that advanced.)

davidmurdoch commented:
@corpulentcoffee `:focus-within` also matches the element itself, not just its children. If the HTML were something like `<div tabindex="0"></div>` it would be focusable and the selector `div:focus-within` would match. I know it doesn't really matter because it detect this case either way, but just thought you'd like to know.

corpulentcoffee commented:
> @corpulentcoffee `:focus-within` also matches the element itself, not just its children. If the HTML were something like `<div tabindex="0"></div>` it would be focusable and the selector div:focus-within would match. I know it doesn't really matter because it detect this case either way, but just thought you'd like to know.

Ah, yes, thanks for the correction! That is indeed something I would like to know.
